compiler/rustc_typeck/src/coherence/orphan.rs

```diff
@@ -330,6 +344,25 @@ fn emit_orphan_check_error<'tcx>(
     })
 }
 
+fn emit_orphan_check_for_impl_on_types<'tcx, T: rustc_errors::EmissionGuarantee>(
+    full_impl_span: Span,
+    self_ty_span: Span,
+    ptr_ty: &ty::TypeAndMut<'_>,
+    diag: &mut DiagnosticBuilder<'tcx, T>,
+) {
+    let mut_key = if ptr_ty.mutbl == rustc_middle::mir::Mutability::Mut { "mut " } else { "" };
+    let msg_sugg =
+        format!("you may want implement it on `{}` or `&{}{}`", ptr_ty.ty, mut_key, ptr_ty.ty);
```

Comment:
This message does not align with the suggestion, which is to introduce a newtype.

Also, be careful about using `ptr_ty` to format the newtype wrapper's name. This will print garbage if we try implementing `impl ForeignTrait for *mut ()` for example.

I wonder if this newtype suggestion could be more general. Right now you're emitting this only if we see `*const/mut T`, so either rename `emit_orphan_check_for_impl_on_types` to be more specific, or find a way to make this more general.